File: crates/ts_encoder/examples/support/codec.rs
//! Frozen codec constructions; no serialization algorithms live in this adapter.
use crate::protocol::{hex, Session};
use serde_json::{json, Value};
use ts_ast::{
    AstBuilder, ContentMapperSourceFileInfo, EagerJsDocProvider, ExternalModuleIndicatorOptions,
    FactoryMethods, FileReference, JsString, MappedDiagnosticDirective, NodeId, NodeKind,
    SourceFileParseOptions, SourceHash, SpanSegment, SyntaxKind,
};
use ts_core::{LanguageVariant, ScriptKind, TextRange};
use ts_jsstring::SourceText;
pub fn scenario(name: &str) -> bool {
    matches!(
        name,
        "literal/StringLiteral"
            | "literal/NumericLiteral"
            | "literal/BigIntLiteral"
            | "literal/RegularExpressionLiteral"
            | "literal/NoSubstitutionTemplateLiteral"
            | "literal/TemplateHead"
            | "literal/TemplateMiddle"
            | "literal/TemplateTail"
            | "literal/Identifier"
            | "literal/PrivateIdentifier"
            | "literal/JsxText"
            | "literal/JSDocText"
            | "literal/JSDocLink"
            | "literal/JSDocLinkPlain"
            | "literal/JSDocLinkCode"
            | "synthetic-expression"
            | "raw-kind"
            | "source-strings"
            | "source-metadata"
            | "source-empty-span"
            | "msgpack-boundaries"
    )
}
fn string(text: &[u8]) -> JsString {
    JsString::from_bytes(text)
}
fn literal(f: &mut AstBuilder, name: &str) -> NodeId {
    let text = b"a\xf0\x9f\x98\x80\xed\xa0\x80\xff";
    let raw = br"raw\n\uD800";
    let flags = -1;
    match name {
        "StringLiteral" => f.new_string_literal(string(text), flags),
        "NumericLiteral" => f.new_numeric_literal(string(text), flags),
        "BigIntLiteral" => f.new_big_int_literal(string(text), flags),
        "RegularExpressionLiteral" => f.new_regular_expression_literal(string(text), flags),
        "NoSubstitutionTemplateLiteral" => {
            f.new_no_substitution_template_literal(string(text), flags)
        }
        "TemplateHead" => f.new_template_head(string(text), string(raw), flags),
        "TemplateMiddle" => f.new_template_middle(string(text), string(raw), flags),
        "TemplateTail" => f.new_template_tail(string(text), string(raw), flags),
        "Identifier" => f.new_identifier(string(text)),
        "PrivateIdentifier" => f.new_private_identifier(string(text)),
        "JsxText" => f.new_jsx_text(string(text), true),
        "JSDocText" => {
            let pieces = f
                .text_slice(vec![string(b"a"), string(b""), string(&text[1..])])
                .unwrap();
            f.new_js_doc_text(pieces)
        }
        "JSDocLink" | "JSDocLinkPlain" | "JSDocLinkCode" => {
            let name_node = f.new_identifier(string(b"name"));
            let pieces = f
                .text_slice(vec![string(b"a"), string(b""), string(&text[1..])])
                .unwrap();
            match name {
                "JSDocLink" => f.new_js_doc_link(Some(name_node), pieces),
                "JSDocLinkPlain" => f.new_js_doc_link_plain(Some(name_node), pieces),
                _ => f.new_js_doc_link_code(Some(name_node), pieces),
            }
        }
        _ => unreachable!("validated literal"),
    }
}
fn options(name: &[u8]) -> SourceFileParseOptions {
    SourceFileParseOptions {
        file_name: string(name),
        ..Default::default()
    }
}
fn source(f: &mut AstBuilder, name: &str) -> NodeId {
    let lit = f.new_string_literal(string("😀".as_bytes()), 1024);
    f.node_mut(lit).unwrap().set_range(TextRange::new(0, 6));
    let a = f.new_identifier(string(b"a"));
    f.node_mut(a).unwrap().set_range(TextRange::new(7, 8));
    let b = f.new_identifier(string(b"a"));
    f.node_mut(b).unwrap().set_range(TextRange::new(9, 10));
    let eof = f.new_token(SyntaxKind::EndOfFile.into());
    f.node_mut(eof).unwrap().set_range(TextRange::new(11, 11));
    let nodes = f.node_slice(vec![Some(lit), Some(a), Some(b)]).unwrap();
    let list = f.new_list(TextRange::new(0, 10), nodes).unwrap();
    let opts = SourceFileParseOptions {
        file_name: string(b"/s06/codec.ts"),
        path: string(b"/s06/codec.ts"),
        external_module_indicator_options: ExternalModuleIndicatorOptions {
            jsx: true,
            force: true,
        },
    };
    let root = f.new_source_file(
        opts,
        SourceText::from_loaded_bytes("'😀' a a\n".as_bytes()),
        Some(list),
        Some(eof),
    );
    f.node_mut(root).unwrap().set_range(TextRange::new(0, 11));
    let nodes = f.node_count();
    let texts = f.text_count();
    {
        let sf = f.source_file_mut(root).unwrap();
        sf.node_count = nodes;
        sf.text_count = texts;
        sf.script_kind = ScriptKind::TS;
        sf.language_variant = LanguageVariant::JSX;
        sf.hash = SourceHash {
            hi: 0x0123_4567_89ab_cdef,
            lo: 0xfedc_ba98_7654_3210,
        };
    }
    if name == "source-metadata" || name == "source-empty-span" {
        let missing_a = f.new_identifier(string(b"missing"));
        let missing_lit = f.new_string_literal(string(b"missing"), 0);
        let one = f.new_source_file(options(b"/s06/one.ts"), SourceText::default(), None, None);
        let two = f.new_source_file(options(b"/s06/two.ts"), SourceText::default(), None, None);
        let canonical = f.new_source_file(
            options(b"/s06/canonical.ts"),
            SourceText::default(),
            None,
            None,
        );
        let module_augmentations = f
            .source_nodes(vec![Some(a), Some(missing_a), None])
            .unwrap();
        let imports = f
            .source_nodes(vec![Some(lit), Some(missing_lit), None])
            .unwrap();
        let ambient_module_names = f
            .source_strings(vec![
                string(b""),
                string(b"repeat"),
                string(b"repeat"),
                string(b"\xff"),
            ])
            .unwrap();
        let referenced_files = f
            .source_references(vec![FileReference {
                loc: TextRange::new(1, 5),
                file_name: string(b"\xff"),
                resolution_mode: -1,
                preserve: true,
            }])
            .unwrap();
        let type_reference_directives = f
            .source_references(vec![FileReference {
                loc: TextRange::new(-1, 128),
                file_name: string(b"types"),
                resolution_mode: 99,
                preserve: false,
            }])
            .unwrap();
        let lib_reference_directives = f
            .source_references(vec![FileReference {
                loc: TextRange::new(0, 11),
                file_name: string(b"lib"),
                resolution_mode: 0,
                preserve: true,
            }])
            .unwrap();
        let segments = if name == "source-empty-span" {
            vec![]
        } else {
            vec![
                SpanSegment {
                    virtual_start: 1,
                    virtual_end: 5,
                    original_start: 1,
                    original_end: 5,
                    kind: 0,
                    features: SpanSegment::FEATURE_ALL,
                },
                SpanSegment {
                    virtual_start: 7,
                    virtual_end: 10,
                    original_start: 5,
                    original_end: 6,
                    kind: 2,
                    features: 0,
                },
            ]
        };
        let supplemental_source_files = f.source_nodes(vec![Some(one), Some(two)]).unwrap();
        let diagnostic_directives = f
            .source_diagnostic_directives(vec![
                MappedDiagnosticDirective {
                    original_range: TextRange::new(1, 5),
                    virtual_range: TextRange::new(1, 5),
                    policy: 1,
                    unused_code: -1,
                    unused_message_text: string(b"not serialized"),
                    source: string(b"not serialized"),
                },
                MappedDiagnosticDirective {
                    original_range: TextRange::new(-1, 128),
                    virtual_range: TextRange::new(-1, 256),
                    policy: 255,
                    unused_code: 65536,
                    unused_message_text: string(b"not serialized"),
                    source: string(b"not serialized"),
                },
            ])
            .unwrap();
        let sf = f.source_file_mut(root).unwrap();
        sf.module_augmentations = module_augmentations;
        sf.imports = imports;
        sf.external_module_indicator = Some(root);
        sf.ambient_module_names = ambient_module_names;
        sf.referenced_files = referenced_files;
        sf.type_reference_directives = type_reference_directives;
        sf.lib_reference_directives = lib_reference_directives;
        sf.set_content_mapper_info(ContentMapperSourceFileInfo {
            content_mapper: string(b"mapper"),
            virtual_file_name: string(b"/s06/codec.virtual.ts"),
            original_text: SourceText::from_loaded_bytes("X😀Y\n".as_bytes()),
            span_map: Some(segments.into()),
            supplemental_source_files,
            canonical_source_file: Some(canonical),
            diagnostic_directives,
            ..Default::default()
        });
    } else if name == "msgpack-boundaries" {
        let mut names = vec![JsString::default(); 65536];
        for (i, n) in [0, 31, 32, 255, 256, 65535, 65536, 1]
            .into_iter()
            .enumerate()
        {
            names[i] = string(&vec![b'x'; n]);
        }
        let positions = [0, 127, 128, 255, 256, 65535, 65536, -1];
        let mut references = Vec::with_capacity(16);
        for i in 0..16 {
            let j = i % 8;
            let p = positions[j];
            references.push(FileReference {
                loc: TextRange::new(p, p),
                file_name: names[j].clone(),
                resolution_mode: p,
                preserve: i % 2 == 0,
            });
        }
        let ambient_module_names = f.source_strings(names).unwrap();
        let referenced_files = f.source_references(references).unwrap();
        let sf = f.source_file_mut(root).unwrap();
        sf.ambient_module_names = ambient_module_names;
        sf.referenced_files = referenced_files;
        sf.type_reference_directives = referenced_files.slice(0..15).unwrap();
    }
    root
}
fn bytes(s: &Session, stage: &str, data: &[u8]) {
    s.observe(stage, "bytes", json!({"length":data.len()}));
    for (i, chunk) in data.chunks(65536).enumerate() {
        s.observe(stage, "chunk", json!({"offset":i*65536,"hex":hex(chunk)}));
    }
}
pub fn execute(s: &Session, r: &Value) {
    let mut encoded = Vec::new();
    if !s.stage("encode", || {
        let mut f = AstBuilder::new(SourceText::default(), &ts_arena::Counters::new());
        let name = r["scenario"].as_str().expect("validated codec");
        let (root, sf) = if let Some(name) = name.strip_prefix("literal/") {
            let root = literal(&mut f, name);
            f.node_mut(root).unwrap().set_range(TextRange::new(-1, 2));
            f.node_mut(root).unwrap().set_flags(0xa5a5_a5a5);
            (root, None)
        } else if name == "synthetic-expression" {
            (f.new_token(SyntaxKind::SyntheticExpression.into()), None)
        } else if name == "raw-kind" {
            let root = f.new_token(NodeKind::from_raw(-1));
            f.node_mut(root)
                .unwrap()
                .set_range(TextRange::new(-1, 2_147_483_647));
            f.node_mut(root).unwrap().set_flags(u32::MAX);
            (root, None)
        } else {
            let root = source(&mut f, name);
            (root, Some(root))
        };
        let mut provider = EagerJsDocProvider::default();
        let result = if let Some(sf) = sf {
            ts_encoder::encode_source_file(f.view(), sf, &mut provider)
        } else {
            ts_encoder::encode_node(f.view(), root, None, &mut provider)
        };
        encoded = result.map_err(|e| e.to_string())?.bytes;
        bytes(s, "encode", &encoded);
        Ok(())
    }) {
        return;
    }
    let mut tree = None;
    if !s.stage("decode", || {
        let decoded = ts_encoder::decode_nodes(&encoded, &ts_arena::Counters::new())
            .map_err(|e| e.to_string())?;
        let root = decoded.root.map(|id| {
            let node = decoded.builder.view().node(id).unwrap();
            json!({"kind":node.kind().raw(),"pos":node.pos(),"end":node.end(),"flags":node.flags()})
        });
        s.observe("decode", "root", json!({"node":root}));
        tree = Some(decoded);
        Ok(())
    }) {
        return;
    }
    let tree = tree.expect("successful decoder");
    if !s.stage("decoded_tree", || {
        super::emit_tree(s, tree.builder.view(), tree.root);
        Ok(())
    }) {
        return;
    }
    s.stage("reencode", || {
        let mut provider = EagerJsDocProvider::default();
        let view = tree.builder.view();
        let result = if tree
            .root
            .is_some_and(|id| view.node(id).unwrap().kind() == SyntaxKind::SourceFile)
        {
            ts_encoder::encode_source_file(view, tree.root.unwrap(), &mut provider)
        } else {
            ts_encoder::encode_optional_node(view, tree.root, None, &mut provider)
        };
        bytes(s, "reencode", &result.map_err(|e| e.to_string())?.bytes);
        Ok(())
    });
}

File: scripts/s06_oracle/codec.go
// Frozen construction scripts; all codec behavior calls the unchanged pin.
package main

import (
	"encoding/hex"
	"github.com/microsoft/TypeScript/tsc/internal/api/encoder"
	"github.com/microsoft/TypeScript/tsc/internal/ast"
	"github.com/microsoft/TypeScript/tsc/internal/core"
	"github.com/microsoft/TypeScript/tsc/internal/parser"
	"github.com/microsoft/TypeScript/tsc/internal/spanmap"
	"github.com/microsoft/TypeScript/tsc/internal/tspath"
	"github.com/zeebo/xxh3"
	"strings"
)

var literalScenarios = []string{"StringLiteral", "NumericLiteral", "BigIntLiteral", "RegularExpressionLiteral", "NoSubstitutionTemplateLiteral", "TemplateHead", "TemplateMiddle", "TemplateTail", "Identifier", "PrivateIdentifier", "JsxText", "JSDocText", "JSDocLink", "JSDocLinkPlain", "JSDocLinkCode"}

func codecScenario(name string) bool {
	for _, literal := range literalScenarios {
		if name == "literal/"+literal {
			return true
		}
	}
	switch name {
	case "synthetic-expression", "raw-kind", "source-strings", "source-metadata", "source-empty-span", "msgpack-boundaries", "parsed-js/typedef", "parsed-js/import":
		return true
	}
	return false
}
func codecLiteral(f *ast.NodeFactory, name string) *ast.Node {
	text := "a😀\xed\xa0\x80\xff"
	raw := `raw\n\uD800`
	flags := ast.TokenFlags(-1)
	switch name {
	case "StringLiteral":
		return f.NewStringLiteral(text, flags)
	case "NumericLiteral":
		return f.NewNumericLiteral(text, flags)
	case "BigIntLiteral":
		return f.NewBigIntLiteral(text, flags)
	case "RegularExpressionLiteral":
		return f.NewRegularExpressionLiteral(text, flags)
	case "NoSubstitutionTemplateLiteral":
		return f.NewNoSubstitutionTemplateLiteral(text, flags)
	case "TemplateHead":
		return f.NewTemplateHead(text, raw, flags)
	case "TemplateMiddle":
		return f.NewTemplateMiddle(text, raw, flags)
	case "TemplateTail":
		return f.NewTemplateTail(text, raw, flags)
	case "Identifier":
		return f.NewIdentifier(text)
	case "PrivateIdentifier":
		return f.NewPrivateIdentifier(text)
	case "JsxText":
		return f.NewJsxText(text, true)
	case "JSDocText":
		return f.NewJSDocText([]string{"a", "", text[1:]})
	case "JSDocLink":
		return f.NewJSDocLink(f.NewIdentifier("name"), []string{"a", "", text[1:]})
	case "JSDocLinkPlain":
		return f.NewJSDocLinkPlain(f.NewIdentifier("name"), []string{"a", "", text[1:]})
	case "JSDocLinkCode":
		return f.NewJSDocLinkCode(f.NewIdentifier("name"), []string{"a", "", text[1:]})
	}
	panic("validated literal missing")
}
func codecSource(f *ast.NodeFactory, name string) *ast.Node {
	text := "'😀' a a\n"
	lit := f.NewStringLiteral("😀", 1024)
	lit.Loc = core.NewTextRange(0, 6)
	a := f.NewIdentifier("a")
	a.Loc = core.NewTextRange(7, 8)
	b := f.NewIdentifier("a")
	b.Loc = core.NewTextRange(9, 10)
	eof := f.NewToken(ast.KindEndOfFile)
	eof.Loc = core.NewTextRange(11, 11)
	list := f.NewNodeList([]*ast.Node{lit, a, b})
	list.Loc = core.NewTextRange(0, 10)
	opts := ast.SourceFileParseOptions{FileName: "/s06/codec.ts", Path: tspath.Path("/s06/codec.ts"), ExternalModuleIndicatorOptions: ast.ExternalModuleIndicatorOptions{JSX: true, Force: true}}
	root := f.NewSourceFile(opts, text, list, eof)
	root.Loc = core.NewTextRange(0, 11)
	sf := root.AsSourceFile()
	sf.NodeCount = f.NodeCount()
	sf.TextCount = f.TextCount()
	sf.ScriptKind = 3
	sf.LanguageVariant = 1
	sf.Hash = xxh3.Uint128{Hi: 0x0123456789abcdef, Lo: 0xfedcba9876543210}
	if name == "source-metadata" || name == "source-empty-span" {
		sf.ModuleAugmentations = []*ast.Node{a, f.NewIdentifier("missing"), nil}
		ast.S06SetImports(sf, []*ast.Node{lit, f.NewStringLiteral("missing", 0), nil})
		sf.ExternalModuleIndicator = root
		sf.AmbientModuleNames = []string{"", "repeat", "repeat", "\xff"}
		sf.ReferencedFiles = []*ast.FileReference{{TextRange: core.NewTextRange(1, 5), FileName: "\xff", ResolutionMode: -1, Preserve: true}}
		sf.TypeReferenceDirectives = []*ast.FileReference{{TextRange: core.NewTextRange(-1, 128), FileName: "types", ResolutionMode: 99}}
		sf.LibReferenceDirectives = []*ast.FileReference{{TextRange: core.NewTextRange(0, 11), FileName: "lib", Preserve: true}}
		other := func(name string) *ast.SourceFile {
			return f.NewSourceFile(ast.SourceFileParseOptions{FileName: name}, "", nil, nil).AsSourceFile()
		}
		segments := []spanmap.Segment{{VirtualStart: 7, VirtualEnd: 10, OriginalStart: 5, OriginalEnd: 6, Kind: 2, Features: 0}, {VirtualStart: 1, VirtualEnd: 5, OriginalStart: 1, OriginalEnd: 5, Kind: 0, Features: spanmap.FeatureAll}}
		if name == "source-empty-span" {
			segments = []spanmap.Segment{}
		}
		sf.SetContentMapperInfo(ast.ContentMapperSourceFileInfo{ContentMapper: "mapper", VirtualFileName: "/s06/codec.virtual.ts", OriginalText: "X😀Y\n", SpanMap: spanmap.New(segments), SupplementalSourceFiles: []*ast.SourceFile{other("/s06/one.ts"), other("/s06/two.ts")}, CanonicalSourceFile: other("/s06/canonical.ts"), DiagnosticDirectives: []ast.MappedDiagnosticDirective{
			{OriginalRange: core.NewTextRange(1, 5), VirtualRange: core.NewTextRange(1, 5), Policy: 1, UnusedCode: -1, UnusedMessageText: "not serialized", Source: "not serialized"},
			{OriginalRange: core.NewTextRange(-1, 128), VirtualRange: core.NewTextRange(-1, 256), Policy: 255, UnusedCode: 65536, UnusedMessageText: "not serialized", Source: "not serialized"},
		}})
	} else if name == "msgpack-boundaries" {
		lengths := []int{0, 31, 32, 255, 256, 65535, 65536, 1}
		sf.AmbientModuleNames = make([]string, 65536)
		for i, n := range lengths {
			sf.AmbientModuleNames[i] = strings.Repeat("x", n)
		}
		positions := []int{0, 127, 128, 255, 256, 65535, 65536, -1}
		for i := 0; i < 16; i++ {
			j := i % 8
			p := positions[j]
			sf.ReferencedFiles = append(sf.ReferencedFiles, &ast.FileReference{TextRange: core.NewTextRange(p, p), FileName: sf.AmbientModuleNames[j], ResolutionMode: core.ResolutionMode(p), Preserve: i%2 == 0})
		}
		sf.TypeReferenceDirectives = sf.ReferencedFiles[:15]
	}
	return root
}
func codecBytes(s *session, stage string, raw []byte) {
	s.observe(stage, "bytes", map[string]any{"length": len(raw)})
	for offset := 0; offset < len(raw); offset += 65536 {
		end := min(offset+65536, len(raw))
		s.observe(stage, "chunk", map[string]any{"offset": offset, "hex": hex.EncodeToString(raw[offset:end])})
	}
}
func executeCodec(s *session, r request) {
	var encoded []byte
	if !s.stage("encode", func() error {
		f := ast.NewNodeFactory(ast.NodeFactoryHooks{})
		name := r.data["scenario"].(string)
		var root *ast.Node
		var sf *ast.SourceFile
		switch {
		case strings.HasPrefix(name, "parsed-js/"):
			text := "/** @typedef {number} Foo */ const x=0;"
			if name == "parsed-js/import" {
				text = "/** @import {Foo} from \"bar\" */ const x=0;"
			}
			sf = parser.ParseSourceFile(ast.SourceFileParseOptions{FileName: "/s06/codec.js", Path: "/s06/codec.js"}, text, core.ScriptKindJS)
			root = sf.AsNode()
		case strings.HasPrefix(name, "literal/"):
			root = codecLiteral(f, name[8:])
			root.Loc = core.NewTextRange(-1, 2)
			root.Flags = 0xa5a5a5a5
		case name == "synthetic-expression":
			root = f.NewToken(ast.KindSyntheticExpression)
		case name == "raw-kind":
			root = f.NewToken(ast.Kind(-1))
			root.Loc = core.NewTextRange(-1, 2147483647)
			root.Flags = 0xffffffff
		default:
			root = codecSource(f, name)
			sf = root.AsSourceFile()
		}
		var err error
		if sf == nil {
			encoded, _, err = encoder.EncodeNode(root, nil)
		} else {
			encoded, _, err = encoder.EncodeSourceFile(sf)
		}
		if err == nil {
			codecBytes(s, "encode", encoded)
		}
		return err
	}) {
		return
	}
	var root *ast.Node
	if !s.stage("decode", func() error {
		var err error
		root, err = encoder.DecodeNodes(encoded)
		if err != nil {
			return err
		}
		var node any
		if root != nil {
			node = map[string]any{"kind": int16(root.Kind), "pos": root.Pos(), "end": root.End(), "flags": uint32(root.Flags)}
		}
		s.observe("decode", "root", map[string]any{"node": node})
		return nil
	}) {
		return
	}
	if !s.stage("decoded_tree", func() error { emitDecoded(s, root); return nil }) {
		return
	}
	s.stage("reencode", func() error {
		var raw []byte
		var err error
		if root != nil && root.Kind == ast.KindSourceFile {
			raw, _, err = encoder.EncodeSourceFile(root.AsSourceFile())
		} else {
			raw, _, err = encoder.EncodeNode(root, nil)
		}
		if err == nil {
			codecBytes(s, "reencode", raw)
		}
		return err
	})
}
